perf_test

```diff
@@ -146,50 +152,66 @@ if __name__ == "__main__":
             inst.chunk_size = chunk_size
             inst.thread_cnt = thread_cnt
             inst.cache_size = cache_size
-            inst.setup()
+            inst.setup(wait=False)
 
             for bs in args.bandwidth:
                 if args.write:
                     print(f"{done:.1f}%", end="\r")
-                    res = run_fio(fio, bs=bs, bw=True, write=True)
+                    res = run_fio(fio, bs=bs, bw=True, write=True,
+                                  rebuild=args.rebuild, md_inst=inst)
                     add_setup_params(res, chunk_size=chunk_size,
                                      thread_cnt=thread_cnt,
                                      cache_size=cache_size)
                     results['BW']['write'].setdefault(bs, [])
                     results['BW']['write'][bs].append(res)
                     done += step
 
+                    if args.rebuild:
+                        inst.setup(wait=False)
```

Comment:
This is what the test does:

```python
if rebuild and md_inst:
    md_inst.degrade(md_inst.devs[0])
    md_inst.recover(md_inst.devs[0])
```

All the extra setups that I've added are because I want to ensure that we're in fact in the middle of a rebuild when the test runs. The alternative would be to set it to rebuild once and then run a bunch of tests, but then I have to track if/when the rebuild has finished. So it's easier to recreate a fresh array and then always do a degrade + recover every time.

Reply:
Why do you need to re-setup? Certainly if the same disk is degraded and then recovered it will restart the sync from the start and it will be in a recovery for a time at least.

If you've already done a setup without a wait() then the recover will certainly fail.... That doesn't sound like a good idea.

You absolutely need to setup without assume-clean and wait for sync. Otherwise recovery is going to fail.

If the wait takes too long, then I'd suggest reducing the size of the array.

Reply:
I did some quick tests:

So degrade() already waits for the last sync to complete in the md.py code, before removing the disk. So it doesn't seem like the wait can be too long seeing you're already waiting.

If you degrade/recover without setting up, it works as excepted. If you disable the wait in degrade() then it will stop the existing recovery and restart it when the next recovery happens. Exactly what we want, without resetting up the array.

Remember, you still definitely have to wait after the first setup.

Also changes like this would make sense:

```diff
diff --git a/md.py b/md.py
index d74e5f6cbc84..3bb86fe55481 100644
--- a/md.py
+++ b/md.py
@@ -356,14 +356,21 @@ class MDInstance:
                                str(n + 1), self.md_dev])
         return n + 1
 
-    def degrade(self, dev):
-        self.wait()
+    def degrade(self, dev=None, nowait=False):
+        if dev is None:
+            dev = self.devs[0]
+
+        if not nowait:
+            self.wait()
         subprocess.check_call([self.mdadm, "--manage", self.md_dev, "--quiet",
                                "--fail", dev])
         subprocess.check_call([self.mdadm, "--manage", self.md_dev, "--quiet",
                                "--remove", dev], stderr=subprocess.DEVNULL)
 
-    def recover(self, dev):
+    def recover(self, dev=None):
+        if dev is None:
+            dev = self.devs[0]
+
         subprocess.check_call([self.mdadm, "--manage", self.md_dev, "--quiet",
                                "--add-spare", dev])
         time.sleep(0.1)

```